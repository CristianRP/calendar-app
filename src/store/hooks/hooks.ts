import { TypedUseSelectorHook, useDispatch } from 'react-redux';
import { AppDispatch, RootState } from '..';
import { useState } from 'react';

export const useAppDispatch: () => AppDispatch = useDispatch;
export const useAppSelector: TypedUseSelectorHook<RootState> = useState;
