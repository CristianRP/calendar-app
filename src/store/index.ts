export * from './auth/authSlice';
export * from './ui/uiSlice';
export * from './calendar/calendarSlice';
export * from './hooks';

export * from './store';
