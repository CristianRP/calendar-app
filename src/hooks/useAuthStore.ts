import { AxiosError } from 'axios';
import { calendarApi } from '../api';
import { User, clearErrorMessage, onChecking, onLogin, onLogout, useAppDispatch, useAppSelector } from '../store';

type AxiosErrorResponse = {
  data: {
    msg?: string;
    errors: [
      {
        msg: string;
      }
    ];
  }
}

export const useAuthStore = () => {
  const { status, user, errorMessage } = useAppSelector(state => state.auth);
  const dispatch = useAppDispatch();

  const startLogin = async({ email, password }: User) => {
    dispatch(onChecking());
    try {
      const { data } = await calendarApi.post('/auth', { email, password });
      localStorage.setItem('token', data.token);
      localStorage.setItem('token-init-date', new Date().getTime().toString());
      dispatch(onLogin({ name: data.name, uid: data.uid }));
    } catch(error) {
      dispatch(onLogout('Not valid credentials'))
      setTimeout(() => {
        dispatch(clearErrorMessage());
      }, 10);
    }
  }

  const startRegister = async({ name, email, password, confirmPassword }: User) => {
    dispatch(onChecking());
    try {
      const { data } = await calendarApi.post('/auth/new', { name, email, password, confirmPassword });
      console.log(data);
    } catch(error) {
      const axiosError = error as AxiosError;
      const { data } = axiosError.response! as AxiosErrorResponse;
      let errorsToShow = ''
      if (data.errors) {
        errorsToShow = `<ul>${Object.values(data.errors).map(error => `<li>${error.msg}</li>`).join('')}</ul>`
      } else {
        errorsToShow = data.msg!;
      }

      dispatch(onLogout(errorsToShow));
      setTimeout(() => {
        dispatch(clearErrorMessage());
      }, 10);
    }
  }

  const checkAuthToken = async() => {
    const token = localStorage.getItem('token');
    if (!token) return dispatch(onLogout(''));

    try {
      const { data } = await calendarApi.get('/auth/renew');
      console.log(data);
      localStorage.setItem('token', data.token);
      localStorage.setItem('token-init-date', new Date().getTime().toString());
      dispatch(onLogin({ name: data.name, uid: data.uid }))
    } catch(error) {
      localStorage.clear();
      dispatch(onLogout(''));
    } 
  }

  const startLogout = () => {
    localStorage.clear();
    dispatch(onLogout(''));
  }

  return {
    //* Properties
    status,
    user,
    errorMessage,

    //* Methods
    startLogin,
    startRegister,
    checkAuthToken,
    startLogout,
  }
}
