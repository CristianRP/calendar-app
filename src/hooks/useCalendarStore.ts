import { TCalendarEvent, onAddNewEvent, onDeleteEvent, onUpdateEvent, setActiveEvent } from '../store';
import { useAppDispatch, useAppSelector } from '../store/hooks'

export const useCalendarStore = () => {
  const dispatch = useAppDispatch();
  const { events, activeEvent } = useAppSelector(state => state.calendar);

  const onSetActiveEvent = (event: TCalendarEvent) => {
    dispatch(setActiveEvent(event));
  }

  const startSavingEvent = async(calendarEvent: TCalendarEvent) => {
    if (calendarEvent._id) {
      dispatch(onUpdateEvent({...calendarEvent}));
    } else {
      dispatch(onAddNewEvent({...calendarEvent, _id: new Date().toISOString() }));
    }
  }

  const startDeletingEvent = () => {
    dispatch(onDeleteEvent());
  }

  return {
    //* Properties
    events,
    activeEvent,
    hasEventSelected: Object.keys(activeEvent).length > 0,

    //* Methods
    onSetActiveEvent,
    startSavingEvent,
    startDeletingEvent,
  }
}
