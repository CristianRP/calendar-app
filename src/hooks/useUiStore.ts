import { useAppDispatch, useAppSelector } from '../store/hooks'
import { onCloseDateModal, onOpenDateModal } from '../store';

export const useUiStore = () => {
  const dispatch = useAppDispatch();

  const { isDateModalOpen } = useAppSelector(state => state.ui);

  const openDateModal = () => {
    dispatch(onOpenDateModal());
  }

  const closeDateModal = () => {
    dispatch(onCloseDateModal());
  }

  return {
    //* Properties
    isDateModalOpen,
    
    //* Methods
    openDateModal,
    closeDateModal,
  }
}
