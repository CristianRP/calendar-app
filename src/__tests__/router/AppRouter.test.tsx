import { render, screen } from '@testing-library/react';
import { AppRouter } from '../../router';
import { useAuthStore } from '../../hooks/useAuthStore';
import { User } from '../../store';
import { MemoryRouter } from 'react-router-dom';

vi.mock('../../hooks/useAuthStore');
vi.mock('../../calendar/pages/CalendarPage', () => ({
  CalendarPage: () => <h1>CalendarPage</h1>
}))

describe('Tests on <AppRouter />', () => {
  const mockCheckAuthToken = vi.fn();

  beforeAll(() => {
    vi.clearAllMocks();
  })

  test('should show the loading page and call checkAuthToken', () => {
    vi.mocked(useAuthStore).mockReturnValue({
      status: 'checking',
      user: {} as User,
      errorMessage: '',
      startLogin: vi.fn(),
      startRegister: vi.fn(),
      checkAuthToken: mockCheckAuthToken,
      startLogout: vi.fn()
    })
    const { container } = render(<AppRouter />);

    expect(container.innerHTML).toBe('<h3>Checking</h3>');
    expect(screen.getByText('Checking')).toBeTruthy();
    expect(mockCheckAuthToken).toHaveBeenCalled();
  });

  test('should show the calendar if user is authenticated', () => {
    vi.mocked(useAuthStore).mockReturnValue({
      status: 'authenticated',
      user: {} as User,
      errorMessage: '',
      startLogin: vi.fn(),
      startRegister: vi.fn(),
      checkAuthToken: mockCheckAuthToken,
      startLogout: vi.fn()
    });

    render(
      <MemoryRouter>
        <AppRouter />
      </MemoryRouter>
    )

    expect(screen.getByText('CalendarPage')).toBeTruthy();
  });
});