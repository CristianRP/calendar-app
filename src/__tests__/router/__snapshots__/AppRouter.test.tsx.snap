// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests on <AppRouter /> > should show the calendar if user is authenticated 1`] = `
<div>
  <div
    class="container mx-auto h-screen flex justify-center items-center"
  >
    <div
      class="flex flex-wrap w-full"
    >
      <div
        class="w-full md:w-1/2 "
      >
        <div
          class="rounded-lg shadow-lg bg-white p-6 h-full flex flex-col justify-center"
        >
          <h3
            class="text-center text-gray-700 text-2xl font-semibold pt-6"
          >
            Login
          </h3>
          <form
            class="p-12"
          >
            <div
              class="mb-4"
            >
              <input
                class="w-full rounded-lg"
                name="loginEmail"
                placeholder="Email"
                type="text"
                value=""
              />
            </div>
            <div
              class="mb-4"
            >
              <input
                class="w-full rounded-lg"
                name="loginPassword"
                placeholder="Password"
                type="password"
                value=""
              />
            </div>
            <div
              class="mb-4"
            >
              <input
                class="w-1/2 mt-2 px-4 py-2 bg-blue-700 text-white rounded-2xl hover:bg-blue-600 focus:outline-none focus:bg-blue-500 focus:ring focus:ring-blue-400 cursor-pointer"
                type="submit"
                value="Login"
              />
            </div>
          </form>
        </div>
      </div>
      <div
        class="w-full md:w-1/2 "
      >
        <div
          class="rounded-lg shadow-2xl p-6 bg-gradient-to-r from-cyan-500 to-blue-500"
        >
          <h3
            class="text-2xl text-white text-center font-semibold pt-6"
          >
            Register
          </h3>
          <form
            class="p-12"
          >
            <div
              class="w-full mb-4"
            >
              <input
                class="w-full rounded-lg"
                id="name"
                name="registerName"
                placeholder="Name"
                type="text"
                value=""
              />
            </div>
            <div
              class="w-full mb-4"
            >
              <input
                class="w-full rounded-lg"
                id="email"
                name="registerEmail"
                placeholder="Email"
                type="email"
                value=""
              />
            </div>
            <div
              class="w-full mb-4"
            >
              <input
                class="w-full rounded-lg ring-1 ring-red-600 focus:ring-1 focus:ring-red-600"
                id="password"
                name="registerPassword"
                placeholder="Password"
                type="password"
                value=""
              />
            </div>
            <div
              class="w-full mb-4"
            >
              <input
                class="w-full rounded-lg ring-1 ring-red-600 focus:ring-1 focus:ring-red-600"
                id="confirm-password"
                name="registerConfirmPassword"
                placeholder="Confirm Password"
                type="password"
                value=""
              />
            </div>
            <div
              class="w-full mb-4"
            >
              <input
                class="w-1/1.5 mt-2 px-4 py-2 bg-white text-blue-700 rounded-2xl hover:bg-slate-100 focus:outline-none focus:bg-white focus:ring focus:ring-blue-400 cursor-pointer"
                type="submit"
                value="Create Account"
              />
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Tests on <AppRouter /> > should show the login if user is not authenticated 1`] = `
<div>
  <div
    class="container mx-auto h-screen flex justify-center items-center"
  >
    <div
      class="flex flex-wrap w-full"
    >
      <div
        class="w-full md:w-1/2 "
      >
        <div
          class="rounded-lg shadow-lg bg-white p-6 h-full flex flex-col justify-center"
        >
          <h3
            class="text-center text-gray-700 text-2xl font-semibold pt-6"
          >
            Login
          </h3>
          <form
            class="p-12"
          >
            <div
              class="mb-4"
            >
              <input
                class="w-full rounded-lg"
                name="loginEmail"
                placeholder="Email"
                type="text"
                value=""
              />
            </div>
            <div
              class="mb-4"
            >
              <input
                class="w-full rounded-lg"
                name="loginPassword"
                placeholder="Password"
                type="password"
                value=""
              />
            </div>
            <div
              class="mb-4"
            >
              <input
                class="w-1/2 mt-2 px-4 py-2 bg-blue-700 text-white rounded-2xl hover:bg-blue-600 focus:outline-none focus:bg-blue-500 focus:ring focus:ring-blue-400 cursor-pointer"
                type="submit"
                value="Login"
              />
            </div>
          </form>
        </div>
      </div>
      <div
        class="w-full md:w-1/2 "
      >
        <div
          class="rounded-lg shadow-2xl p-6 bg-gradient-to-r from-cyan-500 to-blue-500"
        >
          <h3
            class="text-2xl text-white text-center font-semibold pt-6"
          >
            Register
          </h3>
          <form
            class="p-12"
          >
            <div
              class="w-full mb-4"
            >
              <input
                class="w-full rounded-lg"
                id="name"
                name="registerName"
                placeholder="Name"
                type="text"
                value=""
              />
            </div>
            <div
              class="w-full mb-4"
            >
              <input
                class="w-full rounded-lg"
                id="email"
                name="registerEmail"
                placeholder="Email"
                type="email"
                value=""
              />
            </div>
            <div
              class="w-full mb-4"
            >
              <input
                class="w-full rounded-lg ring-1 ring-red-600 focus:ring-1 focus:ring-red-600"
                id="password"
                name="registerPassword"
                placeholder="Password"
                type="password"
                value=""
              />
            </div>
            <div
              class="w-full mb-4"
            >
              <input
                class="w-full rounded-lg ring-1 ring-red-600 focus:ring-1 focus:ring-red-600"
                id="confirm-password"
                name="registerConfirmPassword"
                placeholder="Confirm Password"
                type="password"
                value=""
              />
            </div>
            <div
              class="w-full mb-4"
            >
              <input
                class="w-1/1.5 mt-2 px-4 py-2 bg-white text-blue-700 rounded-2xl hover:bg-slate-100 focus:outline-none focus:bg-white focus:ring focus:ring-blue-400 cursor-pointer"
                type="submit"
                value="Create Account"
              />
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
